### PR TITLE
Issue/566

### LINKED ISSUE #566: Populate the list of trailer headers (if present) in chunked requests
Chunked transfer-encoding may include trailer headers in the request after the last chunk. Http4s seems to parse these correctly but populating the list of them (accessed via `request.trailerHeaders`) seems to be currently not implemented.

Although the usage of trailer headers is not very common, there are some good use cases for them (e.g. passing the checksum of the request stream afterwards). Thus, it would be a useful feature to implement.

### PR DESCRIPTION
**Pull request by bryce-anderson**
closes #566 

I don't like how ugly it became but it does seem pretty self contained.

### CODE REVIEW

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala [hunk 1]
```diff
@@ -220,13 +220,33 @@ private final class Http1Connection(val requestKey: RequestKey,
...
         val (rawBody,_) = collectBodyFromParser(buffer, terminationCondition)
 
+        // to collect the trailers we need a cleanup helper and a Task in the attribute map
+        val (trailerCleanup, attributes) =
+          if (parser.getHttpVersion().minor == 1 && parser.isChunked()) {
+            val trailers = new AtomicReference(Headers.empty)
+
+            val attrs = AttributeMap.empty.put(Message.Keys.TrailerHeaders, Task.suspend {
+              if (parser.contentComplete()) Task.now(trailers.get())
+              else Task.fail(new java.io.IOException("Attempted to collect trailers before the body was complete."))
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala]: Minor: I'd probably call this an IllegalStateException.

**Comment by rossabaker** [comment]
Are you worried about performance or just ugliness?  It's a tiny bit of extra work, but I suspect it's not significant.

Few other servers properly support these, but it's a very handy feature when you need them.  I'm :+1: unless the performance proves ghastly.

**Comment by bryce-anderson** [comment]
I don't think the performance will be worth worrying about. I just wasn't
all that pleased by the junction of `Task` and unchecked effects. I imagine
I can live with it.

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala (hunk 1 above)
Comment (bryce-anderson) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala]: :+1:

**Comment by bryce-anderson** [comment]
I think this is ready. I have a half :+1: so far, I'll merge this tomorrow unless someone beats me to it.

**Comment by rossabaker** [comment]
Here's the other ½ * :+1:

**Pull request merged**